src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +475,94 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
+                        }
+
+                        continue;
+                    }
 
-                        //make sure test case cover post saving (for foreign key update)
-                        if (propertyChange == null)
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        var newPropertyChange = CreateEntityPropertyChange(oldValue, newValue, propertyInfo);
+                        if (propertyChange.NewValue == newPropertyChange.NewValue)
+                        {
+                            // No change
+                            entityChange.PropertyChanges.Remove(propertyChange);
+                        }
+                        else
                         {
-                            if (propertyEntry.HasChanged())
-                            {
-                                var propertyInfo = entityEntry.GetPropertyInfo(property.Name);
-
-                                // Add foreign key
-                                entityChange.PropertyChanges.Add(new EntityPropertyChange
-                                {
-                                    NewValue = propertyEntry.CurrentValue.ToJsonString(),
-                                    OriginalValue = propertyEntry.OriginalValue.ToJsonString(),
-                                    PropertyName = property.Name,
-                                    PropertyTypeFullName = propertyInfo.PropertyType.FullName
-                                });
-                            }
-
-                            continue;
+                            // Update foreign key
+                            propertyChange.NewValue = newPropertyChange.NewValue;
                         }
+                    }
+                }
+
+                /* Update tracked properties */
+                var auditedPropertyNames = entityChange.PropertyChanges
+                                             .Select(pc => pc.PropertyName)
+                                             .Except(foreignKeyPropertyNames)
+                                             .Distinct()
+                                             .ToList();
 
-                        if (propertyChange.OriginalValue == propertyChange.NewValue)
+                foreach(var propertyName in auditedPropertyNames)
+                {
+                    var memberEntry = entityEntry.Member(propertyName);
+                    var primitiveOrComplexProperty = memberEntry is DbPropertyEntry || memberEntry is DbComplexPropertyEntry;
+                    if (!primitiveOrComplexProperty)
+                    {
+                        continue;
+                    }
+
+                    var propertyEntry = entityEntry.Property(propertyName);
+                    var propertyInfo = entityEntry.GetPropertyInfo(propertyName);
+                    var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == propertyName);
```

Comment:
```suggestion
                    var propertyChange = entityChange.PropertyChanges.First(pc => pc.PropertyName == propertyName);
```

`propertyChange` must exist since `propertyName` here is from `entityChange.PropertyChanges`.

Actually, I recommend doing `foreach` over `propertyChanges` instead of doing `.Select(pc => pc.PropertyName)`.

Reply:
> Actually, I recommend doing foreach over propertyChanges instead of doing .Select(pc => pc.PropertyName).

Do you mean do it for `audited properties`? on top of the looping done for `foreign keys` update above?

Reply:
You are already doing it for `audited properties`.

I mean not to select `pc.PropertyName` since you want `pc` anyway.
You are looping `propertyName` of `audited property changes` and then retrieving the `audited property change` again.

Reply:
👍 

Do you know for `foreign keys` updates, why do we add new `PropertyChange` if it does not exist?

Reply:
iirc when you set the reference, e.g. `post.Blog = blog`, EF's `BlogId` is only set when you save changes.

Reply:
If that is the case, we might have issues with adding foreign keys that are not supposed to be tracked.

`entityType.NavigationProperties` should filter by the names in `PropertyChanges`

Reply:
I think you mean it should filter by `ShouldSavePropertyHistory`?

Reply:
Looking at 
[This PR]
```csharp
var foreignKeyPropertyNames = entityType.NavigationProperties
                                                .SelectMany(foreignKeys => foreignKeys.GetDependentProperties())
                                                .Select(p => p.Name)
                                                .Distinct()
                                                .ToList();
```
[Original Code](https://github.com/aspnetboilerplate/aspnetboilerplate/blob/aff7c50e886c330694435a543ce14948500d0da5/src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs#L502-L519)


`foreignKeyPropertyNames` will contain all the foreign keys in the entity type. 
We should filter by
- Name of navigation property which appears in `PropertyChanges`
  - **add/update** the foreign keys of this navigation property in `PropertyChanges`
- For foreign keys whose navigation properties did not appear in `PropertyChanges`
  - **update** the foreign key in `PropertyChanges`

Reply:
I think you mean **remove/update** and **add** respectively?
That's what we're already doing and why we should *not* filter (otherwise we can't **add**), right?

Reply:
I will reply in the together with next flow.